File: web/js/dijkstraneg.js
window.DijkstraNeg = (function () {
  // Binary min-heap storing [distance, vertex] pairs
  function MinHeap() {
    this.data = [];
  }

  MinHeap.prototype.push = function (item) {
    this.data.push(item);
    this._bubbleUp(this.data.length - 1);
  };

  MinHeap.prototype.pop = function () {
    var top = this.data[0];
    var last = this.data.pop();
    if (this.data.length > 0) {
      this.data[0] = last;
      this._sinkDown(0);
    }
    return top;
  };

  MinHeap.prototype.size = function () {
    return this.data.length;
  };

  MinHeap.prototype._bubbleUp = function (i) {
    while (i > 0) {
      var parentIdx = (i - 1) >> 1;
      if (this.data[i][0] < this.data[parentIdx][0]) {
        var tmp = this.data[i];
        this.data[i] = this.data[parentIdx];
        this.data[parentIdx] = tmp;
        i = parentIdx;
      } else {
        break;
      }
    }
  };

  MinHeap.prototype._sinkDown = function (i) {
    var length = this.data.length;
    while (true) {
      var left = 2 * i + 1;
      var right = 2 * i + 2;
      var smallest = i;

      if (left < length && this.data[left][0] < this.data[smallest][0]) {
        smallest = left;
      }
      if (right < length && this.data[right][0] < this.data[smallest][0]) {
        smallest = right;
      }
      if (smallest !== i) {
        var tmp = this.data[i];
        this.data[i] = this.data[smallest];
        this.data[smallest] = tmp;
        i = smallest;
      } else {
        break;
      }
    }
  };

  var INF = 1e9;

  function solve(weightMatrix, source, directed) {
    var n = weightMatrix.length;
    var dist = new Array(n).fill(INF);
    var parent = new Array(n).fill(-1);
    var relaxCount = new Array(n).fill(0);
    var iterations = 0;
    var hasNegativeCycle = false;

    dist[source] = 0;

    var pq = new MinHeap();
    pq.push([0, source]);

    while (pq.size() > 0) {
      var top = pq.pop();
      var d = top[0];
      var u = top[1];
      iterations++;

      // Skip stale entries
      if (d > dist[u]) {
        continue;
      }

      relaxCount[u]++;
      if (relaxCount[u] > n) {
        hasNegativeCycle = true;
        break;
      }

      for (var v = 0; v < n; v++) {
        if (weightMatrix[u][v] === 0) continue;

        // Anti-bounce for undirected graphs: skip going back to parent
        if (!directed && v === parent[u]) continue;

        var newDist = dist[u] + weightMatrix[u][v];
        if (newDist < dist[v]) {
          dist[v] = newDist;
          parent[v] = u;
          pq.push([newDist, v]);
        }
      }
    }

    return {
      dist: dist,
      parent: parent,
      hasNegativeCycle: hasNegativeCycle,
      iterations: iterations
    };
  }

  function reconstructPath(parent, source, target) {
    var n = parent.length;
    if (parent[target] === -1 && target !== source) {
      return [];
    }

    var path = [];
    var cur = target;
    var steps = 0;

    while (cur !== -1 && steps <= n) {
      path.push(cur);
      if (cur === source) break;
      cur = parent[cur];
      steps++;
    }

    // Guard against cycles in parent chain
    if (steps > n) {
      return [];
    }

    if (path[path.length - 1] !== source) {
      return [];
    }

    path.reverse();
    return path;
  }

  return {
    MinHeap: MinHeap,
    solve: solve,
    reconstructPath: reconstructPath
  };
})();

File: dijkstraneg.cpp
#include "dijkstraneg.h"

#include <algorithm>
#include <climits>
#include <queue>

static const int INF = INT_MAX / 2;

DijkstraNegResult DijkstraNeg::solve(
    const std::vector<std::vector<int>>& weightMatrix, int source, bool directed)
{
    const int n = static_cast<int>(weightMatrix.size());
    DijkstraNegResult result;
    result.dist.assign(n, INF);
    result.parent.assign(n, -1);

    if (source < 0 || source >= n)
        return result;

    result.dist[source] = 0;

    // Modified Dijkstra for negative weights:
    // Track how many times each vertex is relaxed. If > n, negative cycle.
    std::vector<int> relaxCount(n, 0);

    // Priority queue: (distance, vertex)
    using PII = std::pair<int, int>;
    std::priority_queue<PII, std::vector<PII>, std::greater<PII>> pq;
    pq.push({0, source});

    while (!pq.empty()) {
        auto [d, u] = pq.top();
        pq.pop();
        ++result.iterations;

        // Skip stale entries
        if (d > result.dist[u])
            continue;

        ++relaxCount[u];
        if (relaxCount[u] > n) {
            result.hasNegativeCycle = true;
            break;
        }

        for (int v = 0; v < n; ++v) {
            if (weightMatrix[u][v] == 0)
                continue;   // no edge

            // For undirected graphs: don't traverse back to the parent vertex.
            // This prevents artificial 2-vertex negative cycles (u→v→u→v...)
            // on undirected edges with negative weights.
            if (!directed && v == result.parent[u])
                continue;

            int newDist = result.dist[u] + weightMatrix[u][v];
            if (newDist < result.dist[v]) {
                result.dist[v] = newDist;
                result.parent[v] = u;
                pq.push({newDist, v});
            }
        }
    }

    return result;
}

std::vector<int> DijkstraNeg::reconstructPath(const std::vector<int>& parent,
                                              int source, int target)
{
    std::vector<int> path;
    if (parent.empty()) return path;

    const int n = static_cast<int>(parent.size());
    if (source < 0 || source >= n || target < 0 || target >= n)
        return path;

    // Check if target is reachable
    if (source != target && parent[target] == -1)
        return path;

    int steps = 0;
    for (int cur = target; cur != -1; cur = parent[cur]) {
        path.push_back(cur);
        if (++steps > n) {
            // Cycle in parent chain (caused by negative cycle) — no valid path
            path.clear();
            return path;
        }
    }

    std::reverse(path.begin(), path.end());
    return path;
}
